File: newsroom/newsroom/lang.py
"""Dil kapısı (M14).

Asteria'nın ürettiği metnin gerçekten Türkçe olduğunu ve kaynak başlığının
olduğu gibi kopyalanmadığını doğrular.

Tasarım notu — neden ölçüm, neden kelime listesi değil:

Eski sistem üç el yapımı kelime listesine (`ENGLISH_MARKERS`, `TURKISH_MARKERS`,
`TURKISH_MORPHOLOGY_RE`) ve İngilizce işaret *sayısına* dayanıyordu. Sayı
uzunluğa bağlı olduğu için uzun Türkçe gövdeler yanlışlıkla reddediliyor, her
yanlış rette koda yeni bir istisna fonksiyonu ekleniyordu
(`body_looks_too_english`, `fact_looks_too_english`).

Buradaki kapı sayı değil **yoğunluk** kullanır; yoğunluk uzunluktan bağımsızdır.
Uzun bir Türkçe metinde geçen "Institute for the Study of War" istisna
gerektirmez, çünkü dört sözcük 150 sözcüklük bir metinde zaten düşük yoğunluktur.

Kısa metne (başlık, description) dil sınıflandırması uygulanmaz. Özel ad yoğun
Türkçe başlıklar — "LinkedIn'in başına Dan Shapero geçti" — hiçbir eşikte
güvenilir sınıflanmaz. Kısa metinde asıl risk zaten farklıdır: kaynak başlığının
çevrilmeden geçmesi. O risk `looks_untranslated()` ile doğrudan ölçülür.

Eşikler `newsroom/tests/corpus` üzerinde ampirik seçildi; gözlenen paylar için
`tests/test_lang.py`.
"""

from __future__ import annotations

from dataclasses import dataclass
import re

from rapidfuzz.fuzz import token_set_ratio

# Gövde kapısı eşikleri.
# Korpusta gözlenen: Türkçe gövdelerde en_density p99=0.029, tr_evidence p01=0.51;
# İngilizce metinlerde en_density p01=0.118, tr_evidence p99≈0.00.
# Aradaki boşluk geniş; eşikler boşluğun ortasına değil, Türkçe tarafa güvenli
# mesafede konuldu (yanlış ret, yanlış kabulden pahalıdır).
BODY_MIN_TURKISH_EVIDENCE = 0.15
BODY_MAX_ENGLISH_DENSITY = 0.06

# Ölçüm için gereken en az sözcük sayısı. Altında karar verilmez.
MIN_MEASURABLE_WORDS = 25

# Kaynak metninden kopyalanmışlık eşiği.
# Korpusta 584 yayının Türkçe başlığı ile İngilizce kaynak başlığı arasındaki
# token_set_ratio: medyan 40, p99 62, gözlenen maksimum 68.
UNTRANSLATED_SIMILARITY = 80

_WORD_RE = re.compile(r"[^\W\d_]+", re.UNICODE)

_TURKISH_CHARS = frozenset("çğıöşüÇĞİÖŞÜ")

# İngilizce fonksiyon sözcükleri. İçerik sözcüğü değil yapı sözcüğü oldukları
# için konu ne olursa olsun İngilizce metinde yüksek, Türkçe metinde özel ad
# öbekleri dışında sıfıra yakın yoğunlukta bulunurlar.
_ENGLISH_FUNCTION_WORDS = frozenset({
    "the", "and", "of", "to", "in", "for", "on", "with", "at", "by", "from",
    "as", "is", "are", "was", "were", "be", "been", "has", "have", "had",
    "will", "would", "can", "could", "that", "this", "these", "those", "it",
    "its", "he", "she", "they", "their", "his", "her", "but", "not", "after",
    "before", "over", "into", "about", "than", "then", "who", "what", "which",
    "said", "says", "more", "most", "new", "up", "out", "if", "how", "why",
})

# Türkçe çekim ve yapım ekleri. Kapsayıcı bir morfoloji çözümlemesi değil,
# yoğunluk sinyali üretmeye yeten bir örneklem.
_PROPER_NOUN_CONNECTORS = r"of|for|the|and|de|del|la|le|van|von|der|des|du|di|da"

# Çok sözcüklü özel ad öbekleri: "Institute for the Study of War",
# "Center for Strategic and International Studies", "The Guardian".
#
# Bu öbeklerin içindeki "for", "the", "of" birer İngilizce yapı sözcüğü değil,
# adın parçasıdır. Ölçümden önce çıkarılmazlarsa, iki kurum adı anan temiz bir
# Türkçe gövde İngilizce sayılabilir. Eski sistem bu durumu ayrı bir istisna
# fonksiyonuyla çözüyordu; burada ölçümün kendisi düzeltiliyor.
_PROPER_NOUN_SPAN_RE = re.compile(
    r"\b[A-ZÇĞİÖŞÜ][\w’']*"
    rf"(?:\s+(?:{_PROPER_NOUN_CONNECTORS})\b|\s+[A-ZÇĞİÖŞÜ][\w’']*)+"
)

_TURKISH_SUFFIX_RE = re.compile(
    r"(?:"
    r"lar[ıi]n|leri[n]?|lar[ıi]|ler[ei]?|"
    r"[dt]a[kn]i|[dt]e[kn]i|"
    r"n[ıiuü]n|[ıiuü]n[ıiuü]|"
    r"[ıiuü]yor|[ae]cak|[ae]ce[gğ]i|"
    r"m[ae]k|m[ae]y[ıi]|"
    r"[ıiuü]ld[ıiuü]|[dt][ıiuü]|[dt][ıiuü]r|"
    r"[ae]r[ae]k|[ıiuü]nc[ae]|"
    r"l[ıiuü][gğ][ıiuü]|l[ıiuü]k"
    r")$"
)


@dataclass(frozen=True, slots=True)
class LanguageMeasurement:
    """Bir metnin ham dil ölçümü. Karar değil, girdi."""

    word_count: int
    english_density: float
    turkish_evidence: float

    @property
    def measurable(self) -> bool:
        return self.word_count >= MIN_MEASURABLE_WORDS


def measure(text: str) -> LanguageMeasurement:
    """Metnin Türkçe ve İngilizce sinyal yoğunluklarını ölçer.

    Türkçe sinyali tam metin üzerinden, İngilizce sinyali ise özel ad öbekleri
    çıkarılmış metin üzerinden hesaplanır.
    """
    words = [word.lower() for word in _WORD_RE.findall(text)]
    if not words:
        return LanguageMeasurement(0, 0.0, 0.0)

    total = len(words)
    without_names = _PROPER_NOUN_SPAN_RE.sub(" ", text)
    english = sum(
        1
        for word in _WORD_RE.findall(without_names.lower())
        if word in _ENGLISH_FUNCTION_WORDS
    )
    turkish_chars = sum(1 for word in words if _TURKISH_CHARS.intersection(word))
    turkish_suffix = sum(1 for word in words if _TURKISH_SUFFIX_RE.search(word))

    return LanguageMeasurement(
        word_count=total,
        english_density=english / total,
        turkish_evidence=(turkish_chars + turkish_suffix) / total,
    )


def body_is_turkish(body: str) -> tuple[bool, str | None]:
    """Haber gövdesinin Türkçe olup olmadığına karar verir.

    Yalnız gövde gibi ölçülebilir uzunluktaki metinler için kullanılır.
    """
    reading = measure(body)

    if not reading.measurable:
        return False, f"gövde ölçülemeyecek kadar kısa ({reading.word_count} sözcük)"

    if reading.english_density > BODY_MAX_ENGLISH_DENSITY:
        return False, (
            f"İngilizce yoğunluğu yüksek "
            f"({reading.english_density:.3f} > {BODY_MAX_ENGLISH_DENSITY})"
        )

    if reading.turkish_evidence < BODY_MIN_TURKISH_EVIDENCE:
        return False, (
            f"Türkçe sinyali zayıf "
            f"({reading.turkish_evidence:.3f} < {BODY_MIN_TURKISH_EVIDENCE})"
        )

    return True, None


def looks_untranslated(text: str, source_text: str) -> tuple[bool, str | None]:
    """Metnin kaynağından çevrilmeden geçirilip geçirilmediğini ölçer.

    Başlık ve description için dil sınıflandırması yerine bu kullanılır:
    kısa metinde asıl risk "Türkçe değil" değil, "kaynak başlığı aynen kalmış".
    """
    if not text.strip() or not source_text.strip():
        return False, None

    similarity = token_set_ratio(text.lower(), source_text.lower())
    if similarity >= UNTRANSLATED_SIMILARITY:
        return True, (
            f"kaynak metnine çok benzer (benzerlik {similarity:.0f} "
            f">= {UNTRANSLATED_SIMILARITY})"
        )
    return False, None

File: worker/src/lang.ts
/* `newsroom/newsroom/lang.py` çevirisi.
 *
 * Eşikler korpustan ampirik seçildi ve orijinal dosyada gerekçeleriyle yazılı.
 * Buradaki iş onları yeniden düşünmek değil, aynı sayıyı üretmek.
 *
 * Düzenli ifadelerde bire bir çeviri yapılamıyor ve bunun nedeni önemli:
 * Python'da `\w` ve `\b` Unicode farkındadır, JavaScript'te ASCII'dir. `\b[A-Z…]`
 * kalıbı JS'e olduğu gibi taşınsaydı "Ürün" gibi bir sözcükte sınır hiç
 * eşleşmezdi, çünkü JS'e göre `Ü` zaten sözcük karakteri değil. Türkçe bir
 * haber sitesinde bu sessizce yanlış ölçüm demek.
 *
 * Bu yüzden `\w` → `[\p{L}\p{N}_]` ve `\b` → geriye bakış ile açıkça yazıldı.
 * Çevirinin tuttuğu `tools/parity-lang.mjs` ile korpusun tamamında ölçülüyor.
 */

import { tokenSetRatio } from "./fuzz.ts";

export const BODY_MIN_TURKISH_EVIDENCE = 0.15;
export const BODY_MAX_ENGLISH_DENSITY = 0.06;
export const MIN_MEASURABLE_WORDS = 25;
export const UNTRANSLATED_SIMILARITY = 80;

/* Python: `[^\W\d_]+` (re.UNICODE) — sözcük karakteri ama rakam ve alt çizgi
 * değil, yani harf. JS karşılığı `\p{L}`; `u` bayrağı olmadan Türkçe harfler
 * dışarıda kalır. */
const WORD_RE = /\p{L}+/gu;

const TURKISH_CHARS = new Set("çğıöşüÇĞİÖŞÜ");

const ENGLISH_FUNCTION_WORDS = new Set([
  "the", "and", "of", "to", "in", "for", "on", "with", "at", "by", "from",
  "as", "is", "are", "was", "were", "be", "been", "has", "have", "had",
  "will", "would", "can", "could", "that", "this", "these", "those", "it",
  "its", "he", "she", "they", "their", "his", "her", "but", "not", "after",
  "before", "over", "into", "about", "than", "then", "who", "what", "which",
  "said", "says", "more", "most", "new", "up", "out", "if", "how", "why",
]);

const PROPER_NOUN_CONNECTORS =
  "of|for|the|and|de|del|la|le|van|von|der|des|du|di|da";

/* Çok sözcüklü özel ad öbekleri. `\b` yerine geriye bakış, `\w` yerine
 * `[\p{L}\p{N}_]` — gerekçesi dosya başında. */
const PROPER_NOUN_SPAN_RE = new RegExp(
  "(?<![\\p{L}\\p{N}_])[A-ZÇĞİÖŞÜ][\\p{L}\\p{N}_’']*" +
    `(?:\\s+(?:${PROPER_NOUN_CONNECTORS})(?![\\p{L}\\p{N}_])` +
    "|\\s+[A-ZÇĞİÖŞÜ][\\p{L}\\p{N}_’']*)+",
  "gu",
);

const TURKISH_SUFFIX_RE = new RegExp(
  "(?:" +
    "lar[ıi]n|leri[n]?|lar[ıi]|ler[ei]?|" +
    "[dt]a[kn]i|[dt]e[kn]i|" +
    "n[ıiuü]n|[ıiuü]n[ıiuü]|" +
    "[ıiuü]yor|[ae]cak|[ae]ce[gğ]i|" +
    "m[ae]k|m[ae]y[ıi]|" +
    "[ıiuü]ld[ıiuü]|[dt][ıiuü]|[dt][ıiuü]r|" +
    "[ae]r[ae]k|[ıiuü]nc[ae]|" +
    "l[ıiuü][gğ][ıiuü]|l[ıiuü]k" +
    ")$",
  "u",
);

export interface LanguageMeasurement {
  wordCount: number;
  englishDensity: number;
  turkishEvidence: number;
}

export function measurable(reading: LanguageMeasurement): boolean {
  return reading.wordCount >= MIN_MEASURABLE_WORDS;
}

function words(text: string): string[] {
  return text.match(WORD_RE) ?? [];
}

/** Türkçe sinyali tam metinden, İngilizce sinyali özel ad öbekleri
 *  çıkarılmış metinden ölçülür. */
export function measure(text: string): LanguageMeasurement {
  const lowered = words(text).map((w) => w.toLowerCase());
  if (lowered.length === 0) {
    return { wordCount: 0, englishDensity: 0, turkishEvidence: 0 };
  }

  const total = lowered.length;
  const withoutNames = text.replace(PROPER_NOUN_SPAN_RE, " ");

  let english = 0;
  for (const word of words(withoutNames)) {
    if (ENGLISH_FUNCTION_WORDS.has(word.toLowerCase())) english += 1;
  }

  let turkishChars = 0;
  let turkishSuffix = 0;
  for (const word of lowered) {
    for (const ch of word) {
      if (TURKISH_CHARS.has(ch)) { turkishChars += 1; break; }
    }
    if (TURKISH_SUFFIX_RE.test(word)) turkishSuffix += 1;
  }

  return {
    wordCount: total,
    englishDensity: english / total,
    turkishEvidence: (turkishChars + turkishSuffix) / total,
  };
}

export function bodyIsTurkish(body: string): { ok: boolean; reason: string | null } {
  const reading = measure(body);

  if (!measurable(reading)) {
    return { ok: false, reason: `gövde ölçülemeyecek kadar kısa (${reading.wordCount} sözcük)` };
  }
  if (reading.englishDensity > BODY_MAX_ENGLISH_DENSITY) {
    return {
      ok: false,
      reason: `İngilizce yoğunluğu yüksek (${reading.englishDensity.toFixed(3)} > ${BODY_MAX_ENGLISH_DENSITY})`,
    };
  }
  if (reading.turkishEvidence < BODY_MIN_TURKISH_EVIDENCE) {
    return {
      ok: false,
      reason: `Türkçe sinyali zayıf (${reading.turkishEvidence.toFixed(3)} < ${BODY_MIN_TURKISH_EVIDENCE})`,
    };
  }
  return { ok: true, reason: null };
}

/** Kısa metinde asıl risk "Türkçe değil" değil, "kaynak başlığı aynen kalmış". */
export function looksUntranslated(
  text: string,
  sourceText: string,
): { untranslated: boolean; detail: string | null } {
  if (!text.trim() || !sourceText.trim()) return { untranslated: false, detail: null };

  const similarity = tokenSetRatio(text.toLowerCase(), sourceText.toLowerCase());
  if (similarity >= UNTRANSLATED_SIMILARITY) {
    return {
      untranslated: true,
      detail: `kaynak metnine çok benzer (benzerlik ${similarity.toFixed(0)} >= ${UNTRANSLATED_SIMILARITY})`,
    };
  }
  return { untranslated: false, detail: null };
}
